config/src/lib.rs

```diff
@@ -56,106 +80,199 @@ pub struct ConfigOption {
-/// 7. Return parsed config
+/// 6. Parse config from file and prepend all options to the command line arguments
+/// 7. Re-parse combined arguments
+/// 8. Call associated function parse_values on config object to create a config object
+/// that is populated with proper values based on the parsed argument
+/// 9. Return the populated config object
 ///
 /// This effectively means that config can be either provided on the command line, or
 /// in a file that is specified via environment variable, with options from the command
 /// line taking precedence over the config file.
-///
-/// * `config` The definition of a config that the matcher will be built from
-/// * `config_file_env` Name of the environment variable to read an extra config file from
-/// * `args` The command line parameters to parse the configuration from (first element will be
-/// ignored, as this is the binary name
-pub fn get_matcher<'a>(
-    config_description: &dyn ConfigDescription,
-    config_file_env: &str,
-    args: Vec<OsString>,
-) -> Result<ArgMatches<'a>, Error> {
-    let configuration = config_description.get_config();
-    let options = &configuration.options;
-
-    let mut matches = App::new(configuration.name)
-        .version(configuration.version)
-        .about(configuration.about);
-
-    for option in options.iter() {
-        matches = matches.arg(
-            Arg::with_name(option.name)
+struct ConfigBuilder {}
+
+impl ConfigBuilder {
+    /// The main function in
+    ///
+    /// * `commandline` The command line parameters to parse the configuration from (first element will be
+    /// ignored, as this is the binary name
+    /// * `config_file_env` Name of the environment variable to read an extra config file from
+    pub fn build<T: Configurable>(
+        commandline: Vec<OsString>,
+        config_file_env: &str,
+    ) -> Result<T, Error> {
+        // Parse commandline according to config definition
+        // -> ParsedValues
+        let description = T::get_config_description();
+
+        //
+        let matcher = ConfigBuilder::create_matcher(&description);
+
+        // Overwrite command line arguments with final arguments to parse
+        // if a config file was specified, all options from that file will be
+        // prepended to the command line arguments
+        let commandline =
+            ConfigBuilder::maybe_combine_arguments(matcher.clone(), commandline, config_file_env);
+
+        // Parse command line
+        let test = commandline.clone().expect("ie");
+        println!("Commandline: {:?}", test);
+        let matcher = matcher.get_matches_from(commandline.expect("trnsi"));
+
+        // Convert results from command line parsing into a HashMap<ConfigOption, Vec<String>>
+        // this is then passed to the actual implementation of the configuration for processing
+        let mut result: HashMap<ConfigOption, Option<Vec<String>>> = HashMap::new();
+        for config_option in description.options.clone() {
```

Comment:
This clone() is not needed when the `create_matcher` method gets proper lifetime annotations.
I'm not suggesting we necessarily do this.

https://users.rust-lang.org/t/borrow-later-used-here-i-dont-understand/49886

Reply:
I think we can leave the clone in this case, shouldn't be too expensive and in case it creates issues down the line we can revisit this.